### ISSUE TITLE
Issue while compiling with CPstream and Cpinternal

### ISSUE THREAD

**Issue by ezazwar1** [Riverside-Software/pct#489]
While compiling with cpstream =1256 and CPintrenal = 1256 the below error is displayed and compilation doesnt happen as it should because i have some programs that have issues and the issue is never showed up 

This happens only when i define numThreads>1 if the numthreads =1 then the process runs fine 

```
[PCTCompile] cpstream set to 1256, but Java doesnt know about it. You may have problems if file names have characters outside of ASCII range.
```

**Comment by gquerret** [comment]
Progress and Java don't use the same codepage names, so a conversion method was created in PCT. Or in fact, two methods were created, one for single thread and one for multiple threads, of course with different code... That doesn't make any sense, I'll refactor the code.

**Comment by gquerret** [comment]
You can try pct.jar from this version: https://ci.rssw.eu/job/PCT/job/PCT-489-GetCharset/
Do not mind the failed unit tests, this comes from internal components being migrated.

**Comment by ezazwar1** [comment]
Works great but still there is something left 

This works only when only one database is connected, But when two or more are connected it fails

**Comment by gquerret** [comment]
I don't see how this could work with one database and not with two or more. Could you setup a small test case so that I can reproduce the issue ?

**Comment by ezazwar1** [comment]
Here is the error i am getting in the build.xml

Meanwhile, while i build a test case here is the error details 

I have defined a DBconnectionSet with two database's connected in that and i have referenced that one in PCTCompile

If the PCT compile is having numThreads= 1 then it works when numThreads>1 below is the errors

```
[PCTCompile] PCTCompile - Progress Code Compiler
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] ** Could not recognise argument: Username~. (301)
[PCTCompile] 0 file(s) compiled

BUILD FAILED
C:\agent\_work\18\s\build.xml:308: The following error occurred while executing this line:
C:\agent\_work\18\s\build.xml:562: connect(-db dbname -S port -ld ldbname -H hostame -U ~'Username~' -P ~'Password~' ) :
```

The one in the where the error is thrown is the second database in the DBconnectionSET


Regards,

**Comment by gquerret** [comment]
A test case would be required to find the problem, but it looks like your username is surrounded by single quotes.

**Comment by ezazwar1** [comment]
Thanks for the TIP looks like the file that i was using import the values had double quotes for username and password 
and in build.xml also the double quotes were there, after removing the quotes it worked


It was hard to guess as it was working on the single thread

**Comment by ezazwar1** [comment]
Dear @gquerret 

There is some bug still in place, The compiler shows no errors and god knows how it compiles the programs
Today when i ran the pipeline with the Numthreads> 1 the compiler never showed me a problem and when we deployed the code in production everything was a mess

Then i removed the numthreads and went to normal process and to my surprise most of the programs around 300+ was using a .i which was not in the propath 
and some of the programs needed to connect to two database's 

The errors only showed up when i didnt use the numthreads 
otherwise it shows that the files compiled successfully 

Regards,

**Comment by gquerret** [comment]
Please provide details about your issue.

**Comment by ezazwar1** [comment]
The issue is that the compiler is not showing any errors in the code, for the Issue that i faced personally is that the include file used in around 300+ programs was not reported when Numthreads> 1 and even the programs that had issue like ambiguous table didnt show up.

Strange thing is that PCT compile when  Numthreads> 1 is generating .r for the programs with issues which should technically not compile at all 

In short the PCTCompile is not throwing errors while compiling programs with errors when Numthreads> 1 but works fine when Numthreads = 1

**Comment by gquerret** [comment]
Execute in verbose mode ( with `ant -v` ) and paste the output with a single thread and multiple threads. You should also include the parts of your build.xml that are relevant for this case. As well as PCT version number.

> Strange thing is that PCT compile when Numthreads> 1 is generating .r for the programs with issues which should technically not compile at all

PCT can't create rcode for procedures that don't compile... Something is probably incorrectly configured.

**Comment by ezazwar1** [comment]
For example perposes i am pasting the output of both 

numthreads=1
```
[PCTCompile]  {TheCompanyName.i}
[PCTCompile] -^
[PCTCompile] ** "TheCompanyName.i" was not found. (293)
[PCTCompile] ** C:\Agent\_work\2\s\winapi\TFilAPI.p Could not understand line 15. (193)
[PCTCompile] ** C:\Agent\_work\2\s\winapi\TFilAPI.p Could not understand line 2. (193)
```

for numthreads>1
```
[PCTCompile] Fileset directory C:\Agent\_work\2\s not found in PROPATH
[PCTCompile] winapi\TFilAPI.p [No r-code]
[PCTCompile] 1 file(s) compiled
```

First one throws error and build fails, the numthreads>1 never throws any error and build suceeds. This also shows that it compiled one file and that is also present in the artifact


But the file that is in artifact throws errors in runtime

**Comment by ezazwar1** [comment]
I ran one more test by removing 
          cpinternal=1256
          cpstream=1256

Then the PCT compile WOrks normally as it should, So the bug i suppose shows up only when 
Numthreads>1 and CPinternal =1256 and Cpstream =1256
Otherwise it works fine without defining CPinternal =1256 and Cpstream =1256

**Comment by gquerret** [comment]
I just ran one of my builds with cpinternal and cpstream set to 1256 (with one and more than one thread), the result was always identical. I can't fix a bug that I can't reproduce ; you can provide a small test case -- example [here](https://github.com/Riverside-Software/pct/tree/master/tests/PCTCompile/test10) -- if you want to see any progress.

Just in case, do you have non-ascii characters in your pathnames ?

**Comment by ezazwar1** [comment]
Dear @gquerret  i somehow managed to replicate the issue, The issue will only work if fileset dir="."



<?xml version="1.0" encoding="utf-8"?>

<project name="Test" default="BuildDone">
	
    <loadproperties srcFile="build.properties" />

    <property name="PCT.JAR" location="PCT.jar" />

    <taskdef resource="PCT.properties"       />
    <typedef resource="types.properties"     />

    <ProgressVersion dlcHome="${dlc_home}" majorVersion="progress.major" minorVersion="progress.minor" patchLevel="progress.patch" fullVersion="progress.full" />

	<target name="BuildDone" >
    <mkdir dir="${Output.Dir}" />
    <mkdir dir="${Output.Test.Dir}" />
    <mkdir dir="${Output.Xref.Dir}" />
    <mkdir dir="${Output.Rcode.Dir}" />

    <PCTCompile 
          destDir="${Output.Rcode.Dir}"
          graphicalMode="true"
          dlcHome="${dlc_home}"
          md5="true"
          minSize="false"
          cpinternal="1256"
          cpstream="1256"
          compileUnderscore="true"
          inputchars="32000"
          token="4000"
          multiCompile="true"
          numThreads="5"
          preprocess="false"
          KeepXref="false"
          stringXref="false"
          preprocessDir=""
          XRefDir="${Output.Xref.Dir}"
          displayFiles="0"
          dirSize="50000">
      <fileset dir=".">
        <include name="**/*.p" />
      </fileset>
     <Profiler enabled="true" coverage="true" outputDir="profiler" />
    </PCTCompile>

	</target>
</project>
[Test (2).zip](https://github.com/Riverside-Software/pct/files/9658877/Test.2.zip)


Just create few directories with the .p in it

**Comment by gquerret** [comment]
> The issue will only work if fileset dir="."

Do you mean that the problem only happens with `fileset dir="."` ? And not with `fileset dir="src"` for example ?

**Comment by gquerret** [comment]
I've executed your build, no problem spotted:
```
C:\Users\GillesQUERRET\Downloads\xxxx
λ ant -DdlcHome=%DLC%
Buildfile: C:\Users\GillesQUERRET\Downloads\xxxx\build.xml

BuildDone:
    [mkdir] Created dir: C:\Users\GillesQUERRET\Downloads\xxxx\Work\Output
    [mkdir] Created dir: C:\Users\GillesQUERRET\Downloads\xxxx\Work\Output\Test
    [mkdir] Created dir: C:\Users\GillesQUERRET\Downloads\xxxx\Work\Output\Xref
    [mkdir] Created dir: C:\Users\GillesQUERRET\Downloads\xxxx\Work\Output\Rcode
[PCTCompile] PCTCompile - Progress Code Compiler
[PCTCompile] cpstream est dÚfini Ó 1256, mais Java ne connait pas cet encodage. Vous pouvez avoir des problÞmes si les noms de fichiers contiennent des caractÞres en dehors de lASCII.
[PCTCompile] 7 fichier(s) compilÚ(s)

BUILD SUCCESSFUL
Total time: 0 seconds

C:\Users\GillesQUERRET\Downloads\xxxx
λ ant -DdlcHome=%DLC%
Buildfile: C:\Users\GillesQUERRET\Downloads\xxxx\build.xml

BuildDone:
[PCTCompile] PCTCompile - Progress Code Compiler
[PCTCompile] cpstream est dÚfini Ó 1256, mais Java ne connait pas cet encodage. Vous pouvez avoir des problÞmes si les noms de fichiers contiennent des caractÞres en dehors de lASCII.
[PCTCompile] 0 fichier(s) compilÚ(s)

BUILD SUCCESSFUL
Total time: 0 seconds

C:\Users\GillesQUERRET\Downloads\xxxx
λ ant -DdlcHome=%DLC%
Buildfile: C:\Users\GillesQUERRET\Downloads\xxxx\build.xml

BuildDone:
[PCTCompile] PCTCompile - Progress Code Compiler
[PCTCompile] cpstream est dÚfini Ó 1256, mais Java ne connait pas cet encodage. Vous pouvez avoir des problÞmes si les noms de fichiers contiennent des caractÞres en dehors de lASCII.
[PCTCompile] 0 fichier(s) compilÚ(s)

BUILD SUCCESSFUL
Total time: 0 seconds
```

**Comment by ezazwar1** [comment]
Cant understand french but i can see that 7 files compiled, which is should not because there is an error in the programs its pointing to a  .I that is not available

**Comment by gquerret** [comment]
I was surprised to see that the .p were empty in my work environment. Then noticed the files in your archive are not empty, and referencing a missing include file, so they were overwritten at some point.
The preprocess and preprocessDir attributes seem to be the problem with multi-thread compilation. If you remove those attributes from the build, then you get the compilation errors. I'll investigate a bit later, but you should now have a working build environment.

**Comment by gquerret** [comment]
And note that `preprocessDir=""` mean that the current directory is the target for preprocessed output, effectively changing the content of each file, so it's not a good idea to use the empty value.

**Comment by gquerret** [comment]
As said in previous comment, it's possible to overwrite source code when fileset dir and preprocessDir are identical. A check has been introduced to prevent that.
Note that this doesn't depend on the codepage or on the number of threads ; preprocessDir also shouldn't be set to the empty string.
